Sources/SwiftOperatorPrecedence/OperatorPrecedence+Folding.swift

```diff
@@ -0,0 +1,467 @@
+  @_spi(Testing)
+  public static func makeBinaryOperationExpr(
+    lhs: ExprSyntax, op: ExprSyntax, rhs: ExprSyntax
+  ) -> ExprSyntax {
+    // The form of the binary operation depends on the operator itself,
+    // which will be one of the unresolved infix operators.
+
+    // An operator such as '+'.
+    if let binaryOperatorExpr = op.as(BinaryOperatorExprSyntax.self) {
+      return ExprSyntax(
+        InfixOperatorExprSyntax(
+          leftOperand: lhs,
+          operatorOperand: ExprSyntax(binaryOperatorExpr),
+          rightOperand: rhs)
+      )
+    }
+
+    // A ternary operator x ? y : z.
+    if let ternaryExpr = op.as(UnresolvedTernaryExprSyntax.self) {
+      return ExprSyntax(
+        TernaryExprSyntax(
+          conditionExpression: lhs,
+          ternaryExpr.unexpectedBeforeQuestionMark,
+          questionMark: ternaryExpr.questionMark,
+          ternaryExpr.unexpectedBetweenQuestionMarkAndFirstChoice,
+          firstChoice: ternaryExpr.firstChoice,
+          ternaryExpr.unexpectedBetweenFirstChoiceAndColonMark,
+          colonMark: ternaryExpr.colonMark,
+          secondChoice: rhs)
+      )
+    }
+
+    // An assignment operator x = y.
+    if let assignExpr = op.as(AssignmentExprSyntax.self) {
+      return ExprSyntax(
+        InfixOperatorExprSyntax(
+          leftOperand: lhs,
+          operatorOperand: ExprSyntax(assignExpr),
+          rightOperand: rhs)
+      )
+    }
+
+    // An "is" type check.
+    if let isExpr = op.as(UnresolvedIsExprSyntax.self) {
+      // FIXME: Do we actually have a guarantee that the right-hand side is a
+      // type expression here?
+      return ExprSyntax(
+        IsExprSyntax(
+          expression: lhs,
+          isExpr.unexpectedBeforeIsTok,
+          isTok: isExpr.isTok,
+          typeName: rhs.as(TypeExprSyntax.self)!.type)
+        )
+    }
+
+    // An "as" cast.
+    if let asExpr = op.as(UnresolvedAsExprSyntax.self) {
+      // FIXME: Do we actually have a guarantee that the right-hand side is a
+      // type expression here?
+      return ExprSyntax(
+        AsExprSyntax(
+          expression: lhs,
+          asExpr.unexpectedBeforeAsTok,
+          asTok: asExpr.asTok,
+          asExpr.unexpectedBetweenAsTokAndQuestionOrExclamationMark,
+          questionOrExclamationMark: asExpr.questionOrExclamationMark,
+          typeName: rhs.as(TypeExprSyntax.self)!.type)
+        )
+    }
+
+    // An arrow expression (->).
+    if let arrowExpr = op.as(ArrowExprSyntax.self) {
+      return ExprSyntax(
+        InfixOperatorExprSyntax(
+          leftOperand: lhs,
+          operatorOperand: ExprSyntax(arrowExpr),
+          rightOperand: rhs)
+      )
+    }
+
+    // FIXME: Fallback that we should never need
+    fatalError("Unknown binary operator")
+  }
+
+  /// Determine the associativity between two precedence groups.
+  private func associativity(
+    firstGroup: PrecedenceGroupName?,
+    firstGroupSyntax: SyntaxProtocol?,
```

Comment:
Similar to https://github.com/apple/swift-syntax/pull/619#discussion_r952216153, this should probably also be named differently.